### PR TITLE
dotCMS/core#23436 Block Editor: Create video block node and insert from dotCMS content

### LINKED ISSUE #23436: Block Editor: Create video block node and insert from dotCMS content
### Is your feature request related to a problem? Please describe.
As a user, I want to insert videos in the block editor from the dotCMS system to my block editor.

### Describe the solution you'd like
1. Open the suggestion
2. Select "Videos"
3. Show the modal with the tabs, same as images
4. List all the videos and filter
5. Click the video to add the video

### Describe alternatives you've considered
For now just this, we have other tasks for add videos from external URL and user file system.

### Acceptance criteria
- User should be able to add videos from dotCMS dotAssets
- 
```[tasklist]
### Tasks
- [ ] https://github.com/dotCMS/core/issues/23920
- [ ] https://github.com/dotCMS/core/issues/23920
```

### LINKED ISSUE #23861: Block Editor: Allow users to insert videos from external sources
### User Story

As a user, I want to be able to search and insert videos from external sources (URL) into the block editor.

### Acceptance Criteria

1. User will enter the URL into the field
2. User can insert the image by clicking the button or tabbing into focus the button and hit enter the keyboard.
3. The image gets inserted in the editor

### Proposed Objective

Core Features

### Proposed Priority

Priority 2 - Important

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

_No response_

### Assumptions & Initiation Needs

_No response_

### Sub-Tasks & Estimates

_No response_

### LINKED ISSUE #23863: Bock Editor: Allow user to insert videos from local files
### User Story

As a user, I want to be able to insert an videos from my local files to the editor.

### How it works?

1. Add tabs to the image panel
2. The first tab will be for the current approach and the second will be for this task, uploading videos from the local files.
3. When user clicks "Upload" it will open the file system window to browse for an image
4. After user select the image we upload as dotAsset and inserted, the same approach that we do when user drops an image in the editor.

### Acceptance Criteria

- User should be able to insert an video from it local files
- Make sure the functionality of adding videos from dotCMS doesn't breaks.

### Proposed Objective

Core Features

### Proposed Priority

Priority 2 - Important

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

_No response_

### Assumptions & Initiation Needs

_No response_

### Sub-Tasks & Estimates

_No response_

### PR DESCRIPTION
**Pull request by rjvelazco**
### What we did here

1. Rename all `dot-image-tabview-form` (and its related components) to `dot-asset-form`.
2. Create the `video` block.
3. Make the `dot-asset-form` reusable to work with `images`, `videos`, and any type of `asset` (future feature).

### Chose a Video from dotCMS**

https://user-images.githubusercontent.com/72418962/215238658-57564bfa-54d3-4fcd-b01c-70b2d5715e83.mov

### Upload your videos

https://user-images.githubusercontent.com/72418962/215238702-ef3463fd-c04e-405a-888e-c486f7f1408c.mov

### External Video (URL)

https://user-images.githubusercontent.com/72418962/215238721-df5b5957-b165-453a-9070-f79f7f59e9b9.mov

### JSON Object for Video Block

**External**

```--json
{
    "type": "video",
    "attrs": {
        "src": "https://www.w3schools.com/html/movie.mp4",
        "mineType": null,
        "width": null,
        "height": null,
        "data": null
    }
}
```

**From dotCMS** 

``` --json

{
    "type": "video",
    "attrs": {
        "src": "/dA/642a2416-5917-4541-b6ec-c7e08eb66e43/asset/mov_bbb.mp4",
        "mineType": null,
        "width": 320,
        "height": 176,
        "data": {
            "hostName": "System Host",
            "modDate": "2023-01-31 17:29:01.492",
            "extension": "mp4",
            "publishDate": "2023-01-31 17:29:01.492",
            "assetContentAsset": "34a1783c-f305-4db5-adaf-0beda2f02f62/asset",
            "assetMetaData": {...},
            "asset": "/dA/34a1783c-f305-4db5-adaf-0beda2f02f62/asset/mov_bbb.mp4",
            "contentTypeIcon": "assessment",
            "language": "en-US"
            ...
        }
    }
}
```

### CODE REVIEW

**Comment by rjvelazco** [comment]
#### Notes

Sonarqube throws the following errors that can not be solved at the moment:

1. Replace this `<i>` tag by `<em>`: However, Primeng use the `<i></i>` tag to set icons on inputs. ([See Docs](https://primefaces.org/primeng/inputtext))

Replace-i-tag-by-em-tag

2. Add `subtitles` and `description` files for this video: Users can only upload a video. We should create another issue to allow users to upload `subtitle` and `description` files for their videos. Then, we can solve this `error`.

add-subtitles-and-description-files-for-this-videos

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/block-editor/src/lib/extensions/action-button/actions-menu.extension.ts [hunk 1]
```diff
@@ -161,9 +161,8 @@ function execCommand({
         horizontalLine: () => {
             editor.chain().deleteRange(range).setHorizontalRule().focus().run();
         },
-        image: () => {
-            editor.commands.toggleImageForm(true);
-        }
+        image: () => editor.commands.openAssetForm({ assetType: 'image' }),
+        video: () => editor.commands.openAssetForm({ assetType: 'video' })
```
[the remark this answers is not in the document]
Comment (rjvelazco) [reply core-web/libs/block-editor/src/lib/extensions/action-button/actions-menu.extension.ts:165]: Great!

File: core-web/libs/block-editor/src/lib/extensions/asset-form/components/dot-upload-asset/dot-upload-asset.component.ts [hunk 2]
```diff
@@ -67,4 +78,19 @@ export class DotUploadAssetComponent {
...
+
+    /**
+     * Set vide File and asset src.
+     *
+     * @private
+     * @param {*} file
+     * @param {*} src
+     * @memberof DotUploadAssetComponent
+     */
+    private setFile(file, src) {
```
[the remark this answers is not in the document]
Comment (rjvelazco) [reply core-web/libs/block-editor/src/lib/extensions/asset-form/components/dot-upload-asset/dot-upload-asset.component.ts:90]: Done!

File: core-web/libs/block-editor/src/lib/extensions/asset-form/components/dot-upload-asset/dot-upload-asset.component.ts [hunk 3]
```diff
@@ -39,9 +49,10 @@ export class DotUploadAssetComponent {
      */
     onSelectFile(event) {
```
[the remark this answers is not in the document]
Comment (rjvelazco) [reply core-web/libs/block-editor/src/lib/extensions/asset-form/components/dot-upload-asset/dot-upload-asset.component.ts:50]: Done!

File: core-web/libs/block-editor/src/lib/extensions/asset-form/components/dot-asset-search/components/dot-asset-card/dot-asset-card.component.html [hunk 4]
```diff
@@ -1,10 +1,15 @@
...
+        <div class="thumbnail-container">
             <img
+                *ngIf="contentlet.hasTitleImage; else icon"
                 [src]="getImage(contentlet.inode)"
                 [alt]="contentlet?.fileName || contentlet?.title"
             />
+            <ng-template #icon>
+                <dot-contentlet-icon [icon]="contentletIcon()" [size]="'72px'">
+                </dot-contentlet-icon>
+            </ng-template>
```
[the remark this answers is not in the document]
Comment (rjvelazco) [reply core-web/libs/block-editor/src/lib/extensions/asset-form/components/dot-asset-search/components/dot-asset-card/dot-asset-card.component.html:12]: For sure, totally agree. Are we going to create a new `issue` for that?

File: core-web/libs/block-editor/src/lib/extensions/asset-form/asset-form.extension.ts [hunk 5]
```diff
@@ -6,15 +6,19 @@ import { ComponentRef, ViewContainerRef } from '@angular/core';
...
+export const BUBBLE_ASSET_FORM_PLUGIN_KEY = new PluginKey('bubble-image-form');
 
 declare module '@tiptap/core' {
     interface Commands<ReturnType> {
-        ImageTabviewForm: {
-            toggleImageForm: (value: boolean) => ReturnType;
+        AssetTabviewForm: {
+            openAssetForm: (data: { assetType: EditorAssetTypes }) => ReturnType;
+            closeAssetForm: () => ReturnType;
+            addAsset: (data: { assetType: EditorAssetTypes; payload }) => ReturnType;
```
[the remark this answers is not in the document]
Comment (rjvelazco) [reply core-web/libs/block-editor/src/lib/extensions/asset-form/asset-form.extension.ts]: This is a great idea. 

We already have a register command called: `openForm` for the `dynamic form plugin`. 

https://github.com/dotCMS/core/blob/6b2a9be4943f043f97a99f48fa822a341cc05d0e/core-web/libs/block-editor/src/lib/extensions/bubble-form/bubble-form.extension.ts#L15-L27

However, I like the idea of having a single global command that is used for all forms and that we only pass it the identifier of the form that we want to open. I am trying to make those changes in this PR: https://github.com/dotCMS/core/pull/23919

Maybe we can create an issue called `Block Editor Refactor` to track a PR?

File: core-web/libs/block-editor/src/lib/extensions/asset-form/components/dot-asset-search/components/dot-asset-card/dot-asset-card.component.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    selector: 'dot-asset-card',
+    templateUrl: './dot-asset-card.component.html',
+    styleUrls: ['./dot-asset-card.component.scss'],
+    changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class DotAssetCardComponent {
+    @Input() contentlet: DotCMSContentlet;
+
+    /**
+     *
```
Comment (fmontes) [inline core-web/libs/block-editor/src/lib/extensions/asset-form/components/dot-asset-search/components/dot-asset-card/dot-asset-card.component.ts:15]: Doc desc.

File: core-web/libs/block-editor/src/lib/extensions/asset-form/components/dot-external-asset/dot-external-asset.component.html [hunk 7]
```diff
@@ -1,14 +1,14 @@
...
+        <label for="url">Insert {{ type }} URL</label>
         <input
             id="url"
             #input
             formControlName="url"
             autocomplete="off"
             type="text"
             pInputText
-            placeholder="https://example.com/image.jpg"
+            placeholder="https://example.com/asset"
```
Comment (fmontes) [inline core-web/libs/block-editor/src/lib/extensions/asset-form/components/dot-external-asset/dot-external-asset.component.html:11]:
```suggestion
            placeholder="https://example.com/asset"
```
```suggestion
            [placeholder]="`https://example.com/${type === "video" ? "video.mp4" : "image.jpg"}`"
```

File: core-web/libs/block-editor/src/lib/extensions/asset-form/components/dot-upload-asset/dot-upload-asset.component.ts [hunk 8]
```diff
@@ -14,34 +21,38 @@ export enum STATUS {
...
 
     /**
      * Set Selected File
      *
-     * @param {*} event
+     * @param {{ files: File[] }} event
      * @memberof DotUploadAssetComponent
      */
-    onSelectFile(event) {
+    onSelectFile(event: { files: File[] }) {
```
Comment (fmontes) [inline core-web/libs/block-editor/src/lib/extensions/asset-form/components/dot-upload-asset/dot-upload-asset.component.ts:50]:
```suggestion
    onSelectFile(files: File[]) {
```

File: core-web/libs/block-editor/src/lib/nodes/video/video.node.ts [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            height: {
+                default: null,
+                parseHTML: (element) => element.getAttribute('height'),
+                renderHTML: (attributes) => ({ height: attributes.height })
+            },
+            data: {
+                default: null,
+                parseHTML: (element) => element.getAttribute('data'),
+                renderHTML: (attributes) => ({ data: attributes.data })
+            }
```
Comment (fmontes) [inline core-web/libs/block-editor/src/lib/nodes/video/video.node.ts:42]: Is this the same or almost the same structure as we do on images?

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/block-editor/src/lib/nodes/video/video.node.ts (hunk 9 above)
Comment (rjvelazco) [reply core-web/libs/block-editor/src/lib/nodes/video/video.node.ts:42]: They are different:

- `image` has the following attributes: `src`, `alt`, `title`, `href` (For images with link), and `data` (Contentlet information).

- `video` has the following attributes: `src`, `mineType `, `width `, `height `, and `data` (Contentlet information).

Comment (rjvelazco) [reply core-web/libs/block-editor/src/lib/nodes/video/video.node.ts:42]: Do we want a specific list of `attributes`?

**Pull request merged**